Create CONTRIBUTING.md guidelines

SugarCaneNS opened the pull request:
It may be time to think about codifying some conventions / guidelines for contributors that we can comfortably agree upon.  It would be easier to code if we knew how to linguistically structure module / function / class / method names.
I would request for review the suggestion that it may be viable to loosely build a contributor's guide upon the foundation of the nvaccess/nvda contributing guidelines, found at https://github.com/nvaccess/nvda/wiki/Contributing/.

Thoughts?

lukaszgo1 commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,56 @@
+# Update the local master.
+git pull
+```
+You should use a separate `topic` branch for each issue. All code should usually be based on the latest commit in the official master branch at the time you start the work unless the code is entirely dependent on the code for another issue. Branches should never be based on the "next" branch.
````
We do not have and never had a next branch. This should be deleted

lukaszgo1 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,56 @@
+## Code Style
+### Encoding
+* Where Python files contain non-ASCII characters, they should be encoded in UTF-8.
+	* There should be no Unicode BOM at the start of the file, as this unfortunately breaks one of the translation tools we use (xgettext). Instead, include this as the first line of the file (only if the file contains non-ASCII characters):
```
This should be reworded -  Gettext is used only in NVDA add-on

lukaszgo1 commented on CONTRIBUTING.md:
````diff
@@ -0,0 +1,56 @@
+* Where Python files contain non-ASCII characters, they should be encoded in UTF-8.
+	* There should be no Unicode BOM at the start of the file, as this unfortunately breaks one of the translation tools we use (xgettext). Instead, include this as the first line of the file (only if the file contains non-ASCII characters):
+```# -*- coding: UTF-8 -*-```
+	* This coding comment must also be included if strings in the code (even strings that aren't translatable) contain escape sequences that produce non-ASCII characters; e.g. `\xff`. This is particularly relevant for braille display drivers. This is due to a gettext bug; see https://github.com/nvaccess/nvda/issues/2592#issuecomment-155299911.
````
This is also not relevant here.

lukaszgo1 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,56 @@
+	* This coding comment must also be included if strings in the code (even strings that aren't translatable) contain escape sequences that produce non-ASCII characters; e.g. `\xff`. This is particularly relevant for braille display drivers. This is due to a gettext bug; see https://github.com/nvaccess/nvda/issues/2592#issuecomment-155299911.
+
+### Indentation
+* Indentation must be done with tabs (one per level), not spaces.
```
This is inconsistent with the rest of the project. While tabs have my personal preference saying that might make someone convert everything to tabs creating huge diffs.
What about the following:
When indenting try to be consistent with the rest of the file that you are working on. In case of newly added files indentation with tab's is preferred.

lukaszgo1 commented:
We should also decide about line endings. At the moment @SugarCaneNS had converted everything except JAWS script to Unix line endings. but the project is not Unix only.

SugarCaneNS commented:
I have done (and intend to commit) the following:
* Removed any reference to the non-existent `next` branch.
* Fixed and re-worded references to gettext.
* Removed the reference to the issue with translatable  strings and the linked issue in nvaccess/nvda.
* updated conventions for indentation.
* Added the following comment regarding line-endings (raw md content follows): ```
### Line endings
Because this project is a package that supports cross-platform integration into many other software packages, line endings for most files (with the exception of the JAWS scripts) should contain Unix-style lf (\\n) line endings.  We have deliberately made the JAWS scripts an exception to the norm, due to JAWS' inability to be able to adequately navigationally interact with lf line endings.  Consequently, any files contained within the JAWS subdirectory (including the readme) should use cr/lf (\\r\\n) line endings.  When using the Windows operating system, whilst committing and pushing with Git, or making a pull request, it is strongly recommended that contributors should use the Git for Windows default setting of `check-out Windows-style line endings, commit Unix-style line endings`.```
By the way, are the escape sequences that I have included really necessary?

SugarCaneNS commented:
Okay, for some reason, Github still interpretted the raw syntax anyway.

lukaszgo1 commented:
NVDA plugin and COM server are also Windows only.  It might make sense to use native line endings for the platform on which it is going to be used.

SugarCaneNS commented:
@lukaszgo1, at the very least, I might do as you have suggested for the NVDA addon.  It would be consistent with the recommendation from NVDA devs anyway.

SugarCaneNS commented:
@lukaszgo1, because of the way I have now setup my Git for Windows, it doesn't really make a difference.  Git simply outputs a warning and automatically converts from cr/lf to lf-only line-endings anyway.
